test/dd/test_package.cpp

```diff
@@ -285,21 +285,86 @@ TEST(DDPackageTest, IdentityTrace) {
   auto dd = std::make_unique<dd::Package<>>(4);
   auto fullTrace = dd->trace(dd->makeIdent(), 4);
 
-  ASSERT_EQ(fullTrace.r, 16.);
+  ASSERT_EQ(fullTrace.r, 1.);
+}
+
+TEST(DDPackageTest, CNotKronTrace) {
+  auto dd = std::make_unique<dd::Package<>>(4);
+  auto cxGate = dd->makeGateDD(dd::X_MAT, 1_pc, 0);
+  auto cxGateKron = dd->kronecker(cxGate, cxGate, 2);
+  auto fullTrace = dd->trace(cxGateKron, 4);
+  ASSERT_EQ(fullTrace, 0.25);
 }
 
 TEST(DDPackageTest, PartialIdentityTrace) {
   auto dd = std::make_unique<dd::Package<>>(2);
   auto tr = dd->partialTrace(dd->makeIdent(), {false, true});
   auto mul = dd->multiply(tr, tr);
-  EXPECT_EQ(dd::RealNumber::val(mul.w.r), 4.0);
+  EXPECT_EQ(dd::RealNumber::val(mul.w.r), 1.);
 }
 
-TEST(DDPackageTest, PartialNonIdentityTrace) {
+TEST(DDPackageTest, PartialSWapMatTrace) {
   auto dd = std::make_unique<dd::Package<>>(2);
+  auto dd2 = std::make_unique<dd::Package<>>(2);
   auto swapGate = dd->makeTwoQubitGateDD(dd::SWAP_MAT, 0, 1);
+  auto swapGate2 = dd2->makeTwoQubitGateDD(dd::SWAP_MAT, 0, 1);
   auto ptr = dd->partialTrace(swapGate, {true, false});
-  EXPECT_EQ(ptr.w * ptr.w, 1.);
+  auto fullTrace = dd->trace(ptr, 1);
+  auto fullTraceOriginal = dd2->trace(swapGate2, 2);
+  EXPECT_EQ(dd::RealNumber::val(ptr.w.r), 0.5);
+  // Check that successively tracing out subsystems is the same as computing the
+  // full trace from the beginning
+  EXPECT_EQ(fullTrace.r, fullTraceOriginal.r);
+}
+
+TEST(DDPackageTest, PartialTraceKeepInnerQubits) {
+  // Check that the partial trace computation is correct when tracing out the
+  // outer qubits only. This test shows that we should avoid storing
+  // non-eliminated nodes in the compute table, as this would prevent their
+  // proper elimination in subsequent trace calls.
+
+  const std::size_t numQubits = 8;
+  auto dd = std::make_unique<dd::Package<>>(numQubits);
+  auto dd2 = std::make_unique<dd::Package<>>(numQubits);
```

Comment:
Same as above goes here. You should be able to reduce the number of packages to one.

Reply:
The issue here is that `partialTrace `and `trace `cannot share the same compute table. Depending on the execution order, one computation will produce incorrect results. For example, if `fullTraceOriginal `is computed first, the partial trace computation will be incorrect because all qubits were eliminated and thus the nodes were stored in the compute table. But then, during the partial trace computation, for the top node, the result from the compute table will be used, even though some lower-level qubits should have been eliminated.

Reply:
I think you can work around that, by only inserting nodes into the compute table if all nodes below that node are eliminated as well. That disambiguates the two use cases. And while it limits the caching for the partial trace a little, it should allow to use both functions within the same package instance.

In the end, it would be very undesirable if a user would get unexpected results because they called the trace method followed by the partial trace. We should make sure that both of them can work in parallel, even if that means that one of them will perform slightly worse.

Reply:
Yes, so the cases are disambiguated now by checking before the lookup and insertion into the table that the lower-level qubits are eliminated as well.